File: src/services/profile.service.ts
import type { Profile } from '@/types';
import { SEED_PROFILES } from '@/mock/seed';

const delay = (min = 200, max = 1000): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, Math.random() * (max - min) + min));

let profiles = [...SEED_PROFILES];

export const profileService = {
  async getAll(): Promise<Profile[]> {
    await delay(300, 800);
    return [...profiles];
  },

  async getById(id: string): Promise<Profile> {
    await delay(100, 400);
    const profile = profiles.find((p) => p.id === id);
    if (!profile) throw new Error(`Profile ${id} not found`);
    return { ...profile };
  },

  async create(
    data: Omit<Profile, 'id' | 'status' | 'latency' | 'updated' | 'trafficUsed' | 'trafficTotal'>,
  ): Promise<Profile> {
    await delay(300, 800);
    const newProfile: Profile = {
      id: `profile-${Date.now()}`,
      status: 'inactive',
      latency: 0,
      updated: new Date().toISOString(),
      ...data,
    };
    profiles = [newProfile, ...profiles];
    return { ...newProfile };
  },

  async update(id: string, data: Partial<Profile>): Promise<Profile> {
    await delay(200, 600);
    const index = profiles.findIndex((p) => p.id === id);
    if (index === -1) throw new Error(`Profile ${id} not found`);
    profiles[index] = { ...profiles[index]!, ...data };
    return { ...profiles[index]! };
  },

  async delete(id: string): Promise<void> {
    await delay(200, 500);
    profiles = profiles.filter((p) => p.id !== id);
  },

  async toggleActive(id: string): Promise<Profile> {
    await delay(150, 400);
    const index = profiles.findIndex((p) => p.id === id);
    if (index === -1) throw new Error(`Profile ${id} not found`);

    if (profiles[index]!.status === 'active') {
      profiles[index] = { ...profiles[index]!, status: 'inactive' };
    } else {
      profiles = profiles.map((p) => ({
        ...p,
        status: p.id === id ? 'active' : ('inactive' as const),
      }));
    }

    return { ...profiles[index]! };
  },
};

File: src-tauri/src/service/profile_service.rs
use uuid::Uuid;

use crate::core::AppState;
use crate::models::{Profile, ProfileStatus, ProfileType};
use crate::utils::AppResult;

/// Request payload for creating a profile.
/// Mirrors the frontend's `Omit<Profile, 'id' | 'status' | 'latency' | 'updated' | 'trafficUsed' | 'trafficTotal'>`.
#[derive(Debug, Clone, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateProfileRequest {
    pub name: String,
    #[serde(rename = "type")]
    pub profile_type: String,
    pub config_url: Option<String>,
}

pub fn get_all(state: &AppState) -> Vec<Profile> {
    state.profiles.read().clone()
}

pub fn get_by_id(state: &AppState, id: &str) -> AppResult<Profile> {
    state
        .profiles
        .read()
        .iter()
        .find(|p| p.id == id)
        .cloned()
        .ok_or_else(|| crate::utils::AppError::NotFound(format!("Profile {}", id)))
}

pub fn create(state: &AppState, data: CreateProfileRequest) -> Profile {
    let profile_type = match data.profile_type.as_str() {
        "Subscription" => ProfileType::Subscription,
        "WireGuard" => ProfileType::WireGuard,
        "VLESS" => ProfileType::Vless,
        "Clash Meta" => ProfileType::ClashMeta,
        _ => ProfileType::Custom,
    };

    let profile = Profile {
        id: format!("profile-{}", Uuid::new_v4()),
        name: data.name,
        profile_type,
        status: ProfileStatus::Inactive,
        latency: 0,
        updated: chrono::Utc::now().to_rfc3339(),
        config_url: data.config_url,
        traffic_used: Some(0),
        traffic_total: None,
    };

    let mut profiles = state.profiles.write();
    profiles.insert(0, profile.clone());

    // Dual-write to database if repository is available
    if let Some(rm) = state.get_resource_manager() {
        let _ = rm.profile_repo.insert(&profile);
    }

    profile
}

pub fn update(state: &AppState, id: &str, data: serde_json::Value) -> AppResult<Profile> {
    let mut profiles = state.profiles.write();
    let index = profiles
        .iter()
        .position(|p| p.id == id)
        .ok_or_else(|| crate::utils::AppError::NotFound(format!("Profile {}", id)))?;

    let mut profile = profiles[index].clone();

    if let Some(name) = data.get("name").and_then(|v| v.as_str()) {
        profile.name = name.to_string();
    }
    if let Some(t) = data.get("type").and_then(|v| v.as_str()) {
        profile.profile_type = match t {
            "Subscription" => ProfileType::Subscription,
            "WireGuard" => ProfileType::WireGuard,
            "VLESS" => ProfileType::Vless,
            "Clash Meta" => ProfileType::ClashMeta,
            _ => ProfileType::Custom,
        };
    }
    if let Some(url) = data.get("configUrl").and_then(|v| v.as_str()) {
        profile.config_url = if url.is_empty() {
            None
        } else {
            Some(url.to_string())
        };
    }
    profile.updated = chrono::Utc::now().to_rfc3339();

    profiles[index] = profile.clone();
    Ok(profile)
}

pub fn delete(state: &AppState, id: &str) -> AppResult<()> {
    let mut profiles = state.profiles.write();
    let len_before = profiles.len();
    profiles.retain(|p| p.id != id);
    if profiles.len() == len_before {
        return Err(crate::utils::AppError::NotFound(format!("Profile {}", id)));
    }
    Ok(())
}

pub fn toggle_active(state: &AppState, id: &str) -> AppResult<Profile> {
    let mut profiles = state.profiles.write();
    let index = profiles
        .iter()
        .position(|p| p.id == id)
        .ok_or_else(|| crate::utils::AppError::NotFound(format!("Profile {}", id)))?;

    let currently_active = matches!(profiles[index].status, ProfileStatus::Active);

    if currently_active {
        profiles[index].status = ProfileStatus::Inactive;
    } else {
        // Deactivate all others, activate this one
        for p in profiles.iter_mut() {
            p.status = ProfileStatus::Inactive;
        }
        profiles[index].status = ProfileStatus::Active;
    }

    Ok(profiles[index].clone())
}
